File: Data Structures & Algorithms/combination-target-sum/submission-3.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        backtrack(ans, new ArrayList<>(), 0, target, 0, nums);
        return ans;
    }

    public void backtrack(List<List<Integer>> ans, List<Integer> current, int sum, int target, int n, int[] nums) {
        if (sum > target) {
            return;
        }
        if (sum == target) {
            ans.add(new ArrayList<>(current));
            return;
        }
        for (int i = n; i < nums.length; i++) {
            current.add(nums[i]);
            backtrack(ans, current, sum + nums[i], target, i, nums);
            current.removeLast();
        }
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-3.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        for (int i = 0; i < nums.size(); i++) {
            int target = -nums[i];
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            vector<vector<int>> curr = twoSum(nums, target, i + 1);
            for (vector<int>& vec : curr) {
                vec.push_back(nums[i]);
                ans.push_back(vec);
            }
        }
        return ans;
    }

    vector<vector<int>> twoSum(vector<int>& nums, int target, int start) {
        vector<vector<int>> ans;
        unordered_map<int, int> map;
        for (int i = start; i < nums.size(); i++) {
            int current = nums[i];
            if (map.count(current)) {
                ans.push_back({nums[map[current]], current});
                while (i + 1 < nums.size() && nums[i] == nums[i + 1]) {
                    i++;
                }
            }
            map[target - current] = i;
        }
        return ans;
    }
};
